lib/watchpool/watchpool.go

```diff
@@ -0,0 +1,169 @@
+package watchpool
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/hashicorp/go-memdb"
+)
+
+// WatchPool is a shared indirection for memdb.WatchSet. It allows multiple
+// goroutines (e.g. blocking RPCs) to watch the same set of chans (i.e. radix
+// tree nodes), with only a single WatchSet.Watch, which may spawn many
+// goroutines for large sets.
+//
+// memdb.WatchSet can watch up to aFew (= 32 currently) chans with no additional
+// goroutine overhead, but beyond that it spawns ceiling(N/32) goroutines. For a
+// watch set containing 2048 radix nodes (e.g. ~682 service instances with
+// checks and nodes) that means each blocking query is running 64 goroutines on
+// top of the one actually serving the request. With hundreds or thousands of
+// clients watching the same things, this quickly adds up and places a lot of
+// load on the server. See https://github.com/hashicorp/consul/issues/4984 for
+// more info.
+//
+// By allowing different blocking queries that are watching the same radix nodes
+// to share a WatchSet.Watch, we can limit that number so in the example above,
+// one client still uses 64 goroutines to watch, but now 1000 clients can all be
+// blocking on that same query and still only 64 goroutines are needed (actually
+// 65 since we add a constant one more here per WatchSet).
+//
+// In the case where only a single RPC is blocked on a specific set of radix
+// nodes, we add no extra goroutines of overhead - the "leader" goroutine
+// processes the Watch just as before inline, if it times out with no change
+// observed, then it passes the leadership to one other watcher if any have come
+// along.
+//
+// To use it, just build your memdb.WatchSet like normal, but instead of calling
+// Watch or WatchCtx directly, call `pool.Watch(ctx, ws)`.
+type WatchPool struct {
+	sync.Mutex
+	watchSets map[uint64]*wsEntry
+}
+
+type wsEntry struct {
+	ws memdb.WatchSet
+	// done is closed by the leader Watch when any chan in the WatchSet is closed.
+	done chan struct{}
+	// leaderDone is NOT closed, but has a message sent on it if the current
+	// leader times out and returns to it's caller. If there are "follower" Watch
+	// calls, one of them will receive the message and take over being the leader.
+	leaderDone chan struct{}
+	// refs is a reference count for how many outstanding Watch calls there are
+	// for this WatchSet. It is incremented and decremented atomically.
+	refs uint32
+}
+
+// Watch performs a shared watch on a memdb.WatchSet. It will deduplicate calls
+// with the same WatchSets (i.e. exactly the same nodes in the state store being
+// watched) and run only one ws.WatchCtx call regardless of the number of
+// concurrent callers.
+func (p *WatchPool) Watch(ctx context.Context, ws memdb.WatchSet) error {
+	key, err := p.wsKey(ws)
+	if err != nil {
+		// This shouldn't ever really happen but if we return immediately then it
```

Comment:
I considered that. I _probably should_ but in practice `wsKey` never returns an error anymore and plumbing a logger through for a case that is never hit is ugly. I might just remove it.

The only reason it did return an error was because I started using `hashstructure` which does before realising it doesn't work in this case. I'll just kill the error entirely for now.